holder.yml

```
@@ -56,60 +56,90 @@ paths:
-      summary: Notifies a holder of an available presentation.
-      operationId: notifyPresentationAvailable
+      summary: Initiates an exchange of information.
+      operationId: initiateExchange
       description:
-        This API is the first part of a holder to holder credential exchange over http.
-        Notifies this server-holder that the client-holder (caller) is prepared to deliver a presentation compliant with the provided Verifiable Presentation Request query. This is the first part of a presentation exchange over this HTTP API. This server-holder MUST return the required 'domain' and 'challenge' parameters for the Presentation. This server-holder MAY reply with an alternate Verifiable Presentation Request query if it requires something different than the one proposed by the client-holder. The client-holder MUST return a Presentation in accordance with the response parameters.
+        A client can use this endpoint to initiate an exchange of a particular
+        type. The client can include HTTP POST information related to the
+        details of exchange it would like to initiate. If the server understands
+        the request, it returns a Verifiable Presentation Request. A request
+        that the server cannot understand results in an error.
+      parameters:
+        - $ref: "./components/parameters/path/ExchangeId.yml"
       requestBody:
-        description: Details the client provides to the server to help it decide if the presentation should be made.
+        description:
+          Information related to the type of exchange the client would like
+          to start.
         content:
           application/json:
             schema:
-              $ref: "#/components/schemas/NotifyPresentationAvailableRequest"
+              anyOf:
+                -
+                  {
+                    "type": "object",
+                    "description": "Data necessary to initiate the exchange."
+                  }
+                -
+                  $ref: "#/components/schemas/NotifyPresentationAvailableRequest"
       responses:
         "200":
-          description: Proceed with presentation
+          description: Proceed with exchange.
           content:
             application/json:
               schema:
-                $ref: "#/components/schemas/NotifyPresentationAvailableResponse"
+                $ref: "#/components/schemas/VerifiablePresentationRequest"
         "400":
-          description: Request for presentation is malformed
+          description: Request is malformed.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/ErrorResponse"
         "501":
-          description: Not implemented
+          description: Service not implemented.
         "500":
-          description: internal error
-  /presentations/submissions:
+          description: Internal server error.
+  /exchanges/{exchange-id}/{transaction-uuid}:
     post:
```

```suggestion
    put:
```

Is this what you're looking for, @mprorock?

---

close.  let me take a stab in the code - might make it clearer

---

This PR has gotten to be a bit of a mess with comments and suggestions, but rescanning through to make sure i didn't miss any key points.  focusing on `holder.yml` for these suggestions

the create a new one of type looks workable, though as @OR13 we may want to explicitly define some types for the `POST` body, I understand the use case for @msporny though, and would like to see if we can ideally satisfy a desire for stronger typing and a more flexible use case

```
/exchanges/{exchange-id}:
    post:
      summary: Creates an exchange as specified by type identified in exchange-id.
      operationId: createExchangeData
```

then onto the `PUT` etc.

```
/exchanges/{exchange-id}/{transaction-uuid}:
    put:
      summary: Updates information related to an existing exchange.
      operationId: updateExchangeData
```

There likely should be defined corresponding getters as well, likely allowing query by example if appropriate, e.g.

```
/exchanges/{exchange-id}/{transaction-uuid}:
    get:
      summary: Gets information related to an existing exchange.
      operationId: getExchangeData 
```

and 

```
/exchanges/{exchange-id}:
    get:
      summary: Gets information related to a exchange(s) identified by the exchange id.
      operationId: updateExchangeData
      parameters:
         - in: path
           name: query
           schema:
             type: any
           required: false    
           description: JSON query by example for filtering exchanges matching the id based on property values
```

and an explicit way to delete and exchange (note, in practice this may be a soft delete for audit purposes)

```
/exchanges/{exchange-id}/{transaction-uuid}:
    delete:
      summary: Deletes an exchange as specified by transaction-uuid.
      operationId: deleteExchangeData 
```

---

> ```
> /exchanges/{exchange-id}:
>     post:
>       summary: Creates an exchange as specified by type identified in exchange-id.
>       operationId: createExchangeData
> ```

I'm good with this, let's do this in this PR.

> then onto the `PUT` etc.
> 
> ```
> /exchanges/{exchange-id}/{transaction-uuid}:
>     put:
>       summary: Updates information related to an existing exchange.
>       operationId: updateExchangeData
> ```

I think I'm good with this, let's do this in this PR. I have to go back to our engineering team to see if they see any issues here.

> There likely should be defined corresponding getters as well, likely allowing query by example if appropriate, e.g.
> 
> ```
> /exchanges/{exchange-id}/{transaction-uuid}:
>     get:
>       summary: Gets information related to an existing exchange.
>       operationId: getExchangeData 
> ```

This one is a bit strange for DB's use cases, but we can probably live with it. The first concern that popped into my mind is one of authentication to do the GET. VP can be used to authn (and is in some of DB's use cases)... but if you can't POST/PUT that, then you need some other form of authn (presumably OAuth2, HTTP Signatures, zcaps, etc.).

I'd like to pursue this in another PR (ideally, raised by @mprorock), but not philosophical objections at this point in time.

> and (edit to note: i put the query in as a path parameter mongo style, but this can be handled other ways as well)

Hrm, what do you mean by this? I think you mean "don't use query parameters, put the query parameter into the path"?

> ```
> /exchanges/{exchange-id}:
>     get:
>       summary: Gets information related to a exchange(s) identified by the exchange id.
>       operationId: updateExchangeData
>       parameters:
>          - in: path
>            name: query
>            schema:
>              type: any
>            required: false    
>            description: JSON query by example for filtering exchanges matching the id based on property values
> ```

Confused about this one, need an example to understand more at depth. Suggest we do this one in a separate PR.

> and an explicit way to delete and exchange (note, in practice this may be a soft delete for audit purposes)
> 
> ```
> /exchanges/{exchange-id}/{transaction-uuid}:
>     delete:
>       summary: Deletes an exchange as specified by transaction-uuid.
>       operationId: deleteExchangeData 
> ```

No philosophical objections to this one. Let's do this in another PR.

---

@mprorock in 1 sentence what does `{transaction-uuid}` mean to you?

---

the {transaction-uuid} actual unique identifier that represents the item in question - in this case a specific transaction as a part of a particular exchange

---

> This PR has gotten to be a bit of a mess with comments and suggestions, but rescanning through to make sure i didn't miss any key points. focusing on `holder.yml` for these suggestions
> 
> the create a new one of type looks workable, though as @OR13 we may want to explicitly define some types for the `POST` body, I understand the use case for @msporny though, and would like to see if we can ideally satisfy a desire for stronger typing and a more flexible use case
> 
> ```
> /exchanges/{exchange-id}:
>     post:
>       summary: Creates an exchange as specified by type identified in exchange-id.
>       operationId: createExchangeData
> ```
> 
> then onto the `PUT` etc.
> 
> ```
> /exchanges/{exchange-id}/{transaction-uuid}:
>     put:
>       summary: Updates information related to an existing exchange.
>       operationId: updateExchangeData
> ```
> 
> There likely should be defined corresponding getters as well, likely allowing query by example if appropriate, e.g.
> 
> ```
> /exchanges/{exchange-id}/{transaction-uuid}:
>     get:
>       summary: Gets information related to an existing exchange.
>       operationId: getExchangeData 
> ```
> 
> and (edit to note: i put the query in as a path parameter mongo style, but this can be handled other ways as well)
> 
> ```
> /exchanges/{exchange-id}:
>     get:
>       summary: Gets information related to a exchange(s) identified by the exchange id.
>       operationId: updateExchangeData
>       parameters:
>          - in: path
>            name: query
>            schema:
>              type: any
>            required: false    
>            description: JSON query by example for filtering exchanges matching the id based on property values
> ```
> 
> and an explicit way to delete and exchange (note, in practice this may be a soft delete for audit purposes)
> 
> ```
> /exchanges/{exchange-id}/{transaction-uuid}:
>     delete:
>       summary: Deletes an exchange as specified by transaction-uuid.
>       operationId: deleteExchangeData 
> ```

This is exactly what i had envisioned

---

@OR13 `exchange id` = type of exchange for example TransferOfOilCredential
`transaction-uuid` = instance of transferOfOilCredential Exchange

---

ahh to exchange id => definition, transaction uuid => instance... and both are current communicated via post body arguments... and this PR copies them into URL parameters... I assume we would error if they differed?